src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/PosixSignalRegistration.Unix.cs

```diff
@@ -8,200 +8,128 @@ namespace System.Runtime.InteropServices
+            // on work triggered from the signal handling thread.
+            switch (signal)
             {
-                if (s_registrations.TryGetValue(signo, out List<WeakReference<PosixSignalRegistration>?>? signalRegistrations))
-                {
-                    if (signalRegistrations.Count != 0)
+                case PosixSignal.SIGINT:
+                case PosixSignal.SIGQUIT:
+                case PosixSignal.SIGTERM:
+                    // For terminate/interrupt signals we use a dedicated Thread in case the ThreadPool is saturated.
+                    new Thread(HandleSignal)
                     {
-                        var registrations = new PosixSignalRegistration?[signalRegistrations.Count];
-                        bool hasRegistrations = false;
-                        bool pruneWeakReferences = false;
-
-                        for (int i = 0; i < signalRegistrations.Count; i++)
-                        {
-                            if (signalRegistrations[i]!.TryGetTarget(out PosixSignalRegistration? registration))
-                            {
-                                registrations[i] = registration;
-                                hasRegistrations = true;
-                            }
-                            else
-                            {
-                                // WeakReference no longer holds an object. PosixSignalRegistration got finalized.
-                                signalRegistrations[i] = null;
-                                pruneWeakReferences = true;
-                            }
-                        }
-
-                        if (pruneWeakReferences)
-                        {
-                            signalRegistrations.RemoveAll(item => item is null);
-                        }
+                        IsBackground = true,
+                        Name = ".NET Signal Handler"
+                    }.UnsafeStart((signo, tokens));
+                    break;
 
-                        if (hasRegistrations)
-                        {
-                            return registrations;
-                        }
-                        else
-                        {
-                            Interop.Sys.DisablePosixSignalHandling(signo);
-                        }
-                    }
-                }
-                return null;
+                default:
+                    ThreadPool.UnsafeQueueUserWorkItem(HandleSignal, (signo, tokens));
+                    break;
             }
-        }
 
-        private static void HandleSignal(object? state)
-        {
-            HandleSignal(((int, PosixSignalRegistration?[]))state!);
-        }
+            return 1;
 
-        private static void HandleSignal((int signo, PosixSignalRegistration?[]? registrations) state)
-        {
-            do
+            static void HandleSignal(object? state)
             {
+                (int signo, Token[]? tokens) = ((int, Token[]?))state!;
```

Comment:
There is some more that can be removed here:
`tokens` can no longer be `null`.
`handlersCalled` can be removed (it will always be `true`).
The second argument of `HandleNonCanceledPosixSignal` can also be removed.